File: Tests/UnitTestImpromptuInterface.VBNet/ImpromptuDynamic.vb
Option Strict Off

Imports System.Collections
Imports System.Collections.Generic
Imports System.Dynamic
Imports System.Linq
Imports System.Runtime.CompilerServices
Imports System.Text
Imports NUnit.Framework
Imports ImpromptuInterface
Imports ImpromptuInterface.Dynamic

Namespace VBNET

    <TestFixture()> _
    Public Class ImpromptuDynamic
        Inherits Helper

        <Test()> _
        Public Sub DictionaryPropertyTest()

            Dim tNew As Object = New ImpromptuDictionary()
            tNew.Prop1 = "Test"
            tNew.Prop2 = 42L
            tNew.Prop3 = Guid.NewGuid()

            Dim tActsLike As ISimpeleClassProps = Impromptu.ActLike(Of ISimpeleClassProps)(tNew)

            Assert.AreEqual(tNew.Prop1, tActsLike.Prop1)
            Assert.AreEqual(tNew.Prop2, tActsLike.Prop2)
            Assert.AreEqual(tNew.Prop3, tActsLike.Prop3)
        End Sub

        <Test()> _
        Public Sub DictionaryNullPropertyTest()

            Dim tNew As Object = New ImpromptuDictionary()


            Dim tActsLike As ISimpeleClassProps = Impromptu.ActLike(Of ISimpeleClassProps)(tNew)

            Assert.AreEqual(Nothing, tActsLike.Prop1)
            Assert.AreEqual(0, tActsLike.Prop2)
            Assert.AreEqual(New Guid(), tActsLike.Prop3)
        End Sub

        <Test()> _
        Public Sub GetterAnonTest()
            Dim tAnon = New With { _
             Key .Prop1 = "Test", _
             Key .Prop2 = 42L, _
             Key .Prop3 = Guid.NewGuid() _
            }

            Dim tTest As Object = New ImpromptuGet(tAnon)

            Assert.AreEqual(tAnon.Prop1, tTest.Prop1)
            Assert.AreEqual(tAnon.Prop2, tTest.Prop2)
            Assert.AreEqual(tAnon.Prop3, tTest.Prop3)
        End Sub

        <Test()> _
        Public Sub GetterVoidTest()
            Dim tPoco = New VoidMethodPoco()

            Dim tTest As Object = New ImpromptuGet(tPoco)

            tTest.Action()
        End Sub

        <Test()> _
        Public Sub GetterArrayTest()


            Dim tArray = New Integer() {1, 2, 3}

            Dim tTest As IStringIntIndexer = ImpromptuGet.Create(Of IStringIntIndexer)(tArray)

            Assert.AreEqual(tArray(2).ToString(), tTest(2))
        End Sub






        <Test()> _
        Public Sub GetterObjectTest()
            Dim tNew As Object = New ExpandoObject()
            tNew.Prop1 = "Test"
            tNew.Prop2 = 42L
            tNew.Prop3 = Guid.NewGuid()

            Dim tTest As Object = New ImpromptuGet(tNew)


            Assert.AreEqual(tNew.Prop1, tTest.Prop1)
            Assert.AreEqual(tNew.Prop2, tTest.Prop2)
            Assert.AreEqual(tNew.Prop3, tTest.Prop3)
        End Sub

        <Test()> _
        Public Sub ForwardAnonTest()
            Dim tAnon = New With { _
             Key .Prop1 = "Test", _
             Key .Prop2 = 42L, _
             Key .Prop3 = Guid.NewGuid() _
            }

            Dim tTest As Object = New TestForwarder(tAnon)

            Assert.AreEqual(tAnon.Prop1, tTest.Prop1)
            Assert.AreEqual(tAnon.Prop2, tTest.Prop2)
            Assert.AreEqual(tAnon.Prop3, tTest.Prop3)
        End Sub

        <Test()> _
        Public Sub ForwardVoidTest()
            Dim tPoco = New VoidMethodPoco()

            Dim tTest As Object = New TestForwarder(tPoco)

            tTest.Action()
        End Sub



        <Test()> _
        Public Sub ForwardObjectTest()
            Dim tNew As Object = New ExpandoObject()
            tNew.Prop1 = "Test"
            tNew.Prop2 = 42L
            tNew.Prop3 = Guid.NewGuid()

            Dim tTest As Object = New TestForwarder(tNew)


            Assert.AreEqual(tNew.Prop1, tTest.Prop1)
            Assert.AreEqual(tNew.Prop2, tTest.Prop2)
            Assert.AreEqual(tNew.Prop3, tTest.Prop3)
        End Sub

        <Test()> _
        Public Sub DictionaryMethodsTest()

            Dim tNew As Object = New ImpromptuDictionary()
            tNew.Action1 = New Action(AddressOf Assert.Fail)
            tNew.Action2 = New Action(Of Boolean)(AddressOf Assert.IsFalse)
            tNew.Action3 = New Func(Of String)(Function() "test")
            tNew.Action4 = New Func(Of Integer, String)(Function(arg) "test" & Convert.ToString(arg))


            Dim tActsLike As ISimpeleClassMeth2 = Impromptu.ActLike(Of ISimpeleClassMeth2)(tNew)



            AssertException(Of AssertionException)(AddressOf tActsLike.Action1)

            Assert.AreEqual("test", tActsLike.Action3())

            Assert.AreEqual("test4", tActsLike.Action4(4))
        End Sub

        <Test()> _
        Public Sub ForwardMethodsTest()

            Dim tNew As Object = New ImpromptuDictionary()
            tNew.Action1 = New Action(AddressOf Assert.Fail)
            tNew.Action2 = New Action(Of Boolean)(AddressOf Assert.IsFalse)
            tNew.Action3 = New Func(Of String)(Function() "test")
            tNew.Action4 = New Func(Of Integer, String)(Function(arg) "test" & Convert.ToString(arg))


            Dim tFwd As Object = New TestForwarder(tNew)



            AssertException(Of AssertionException)(Function() tFwd.Action1())
            AssertException(Of AssertionException)(Function() tFwd.Action2(True))

            Assert.AreEqual("test", tFwd.Action3())

            Assert.AreEqual("test4", tFwd.Action4(4))
        End Sub



        Private Shared Function TestOut(dummy As CallSite, [in] As Object, ByRef out As String) As Object
            out = TryCast([in], String)

            Return out IsNot Nothing
        End Function


        <Test()> _
        Public Sub DictionaryMethodsTestWithPropertyAccess()

            Dim tNew As Object = New ImpromptuDictionary()
            tNew.PropCat = "Cat-"
            tNew.Action1 = New Action(AddressOf Assert.Fail)
            tNew.Action2 = New Action(Of Boolean)(AddressOf Assert.IsFalse)
            tNew.Action3 = New ThisFunc(Of String)(Function(this) Convert.ToString(this.PropCat) & "test")

            Dim tActsLike As ISimpeleClassMeth = Impromptu.ActLike(Of ISimpeleClassMeth)(tNew)



            AssertException(Of AssertionException)(AddressOf tActsLike.Action1)

            Assert.AreEqual("Cat-test", tActsLike.Action3())


        End Sub

        <Test()> _
        Public Sub DictionaryNullMethodsTest()

            Dim tNew As Object = New ImpromptuDictionary()

            Dim tActsLike As ISimpleStringMethod = Impromptu.ActLike(Of ISimpleStringMethod)(tNew)

            Assert.AreEqual(False, tActsLike.StartsWith("Te"))



        End Sub


        <Test()> _
        Public Sub ObjectDictionaryWrappedTest()

            Dim tDictionary = New Dictionary(Of String, Object)() From { _
             {"Test1", 1}, _
             {"Test2", 2}, _
             {"TestD", New Dictionary(Of String, Object)() From { _
              {"TestA", "A"}, _
              {"TestB", "B"} _
             }} _
            }

            Dim tNew As Object = New ImpromptuDictionary(tDictionary)

            Assert.AreEqual(1, tNew.Test1)
            Assert.AreEqual(2, tNew.Test2)
            Assert.AreEqual("A", tNew.TestD.TestA)
            Assert.AreEqual("B", tNew.TestD.TestB)
        End Sub

        <Test()> _
        Public Sub InterfaceDictionaryWrappedTest()

            Dim tDictionary = New Dictionary(Of String, Object)() From { _
             {"Test1", 1}, _
             {"Test2", 2L}, _
             {"TestD", New Dictionary(Of String, Object)() From { _
              {"TestA", "A"}, _
              {"TestB", "B"} _
             }} _
            }

            Dim tObject As Object = ImpromptuDictionary.Create(Of IDynamicDict)(tDictionary)
            Dim tNotObject As Object = ImpromptuDictionary.Create(Of INonDynamicDict)(tDictionary)

            Assert.AreEqual(tObject, tNotObject)

            Assert.AreEqual(1, tObject.Test1)
            Assert.AreEqual(2L, tObject.Test2)
            Assert.AreEqual("A", tObject.TestD.TestA)
            Assert.AreEqual("B", tObject.TestD.TestB)

            Assert.AreEqual(1, tNotObject.Test1)
            Assert.AreEqual(2L, tNotObject.Test2)


            Assert.AreEqual(GetType(Dictionary(Of String, Object)), tNotObject.TestD.[GetType]())
            Assert.AreEqual(GetType(ImpromptuDictionary), tObject.TestD.[GetType]())
        End Sub

        <Test()> _
        Public Sub ObjectObjectEqualsTest()
            Dim tDictionary = New Dictionary(Of String, Object)() From { _
             {"Test1", 1}, _
             {"Test2", 2}, _
             {"TestD", New Dictionary(Of String, Object)() From { _
              {"TestA", "A"}, _
              {"TestB", "B"} _
             }} _
            }

            Dim tObject As Object = ImpromptuDictionary.Create(Of IDynamicDict)(tDictionary)
            Dim tNotObject As Object = ImpromptuDictionary.Create(Of INonDynamicDict)(tDictionary)

            Assert.AreEqual(tObject, tNotObject)

            Assert.AreEqual(tObject, tDictionary)

            Assert.AreEqual(tNotObject, tDictionary)
        End Sub

        <Test()> _
        Public Sub ObjectAnnonymousWrapper()
            Dim tData = New Dictionary(Of Integer, String)() From { _
             {1, "test"} _
            }
            Dim tDyn = ImpromptuGet.Create(New With { _
             Key .Test1 = 1, _
             Key .Test2 = "2", _
             Key .IsGreaterThan5 = [Return](Of Boolean).Arguments(Of Integer)(Function(it) it > 5), _
             Key .ClearData = ReturnVoid.Arguments(Function() tData.Clear()) _
            })

            Assert.AreEqual(1, tDyn.Test1)
            Assert.AreEqual("2", tDyn.Test2)
            Assert.AreEqual(True, tDyn.IsGreaterThan5(6))
            Assert.AreEqual(False, tDyn.IsGreaterThan5(4))

            Assert.AreEqual(1, tData.Count)
            tDyn.ClearData()
            Assert.AreEqual(0, tData.Count)

        End Sub

        <Test()> _
        Public Sub TestAnonInterface()
            Dim tInterface As ICollection = ImpromptuGet.Create(Of ICollection)(New With { _
             .CopyArray = ReturnVoid.Arguments(Of Array, Integer)(Function(ar, i) Enumerable.Range(1, 10)), _
             .Count = 10, _
             .IsSynchronized = False, _
             .SyncRoot = Me, _
             .GetEnumerator = [Return](Of IEnumerator).Arguments(Function() Enumerable.Range(1, 10).GetEnumerator()) _
            })

            Assert.AreEqual(10, tInterface.Count)
            Assert.AreEqual(False, tInterface.IsSynchronized)
            Assert.AreEqual(Me, tInterface.SyncRoot)
            Assert.AreEqual(True, tInterface.GetEnumerator().MoveNext())
        End Sub

        <Test()> _
        Public Sub TestBuilder()

            Dim NewD As Object = New ImpromptuBuilder(Of ExpandoObject)()


            Dim tExpandoNamedTest = NewD.Robot(LeftArm:="Rise", RightArm:="Clamp")

            Assert.AreEqual("Rise", tExpandoNamedTest.LeftArm)
            Assert.AreEqual("Clamp", tExpandoNamedTest.RightArm)
        End Sub

        <Test()> _
        Public Sub TestSetupOtherTypes()
            Dim [New] = Builder.[New]().Setup(Expando:=GetType(ExpandoObject), Dict:=GetType(ImpromptuDictionary))

            Dim tExpando = [New].Expando(LeftArm:="Rise", RightArm:="Clamp")

            Dim tDict = [New].Dict(LeftArm:="RiseD", RightArm:="ClampD")

            Assert.AreEqual("Rise", tExpando.LeftArm)
            Assert.AreEqual("Clamp", tExpando.RightArm)
            Assert.AreEqual(GetType(ExpandoObject), tExpando.[GetType]())

            Assert.AreEqual("RiseD", tDict.LeftArm)
            Assert.AreEqual("ClampD", tDict.RightArm)
            Assert.AreEqual(GetType(ImpromptuDictionary), tDict.[GetType]())

        End Sub


        'This test data is modified from MS-PL Clay project http://clay.codeplex.com
        <Test()> _
        Public Sub TestClayFactorySyntax()
            Dim [New] As Object = Builder.[New]()

            If True Then
                Dim person = [New].Person()
                person.FirstName = "Louis"
                person.LastName = "Dejardin"
                Assert.AreEqual("Louis", person.FirstName)
                Assert.AreEqual("Dejardin", person.LastName)
            End If
            If True Then
                Dim person = [New].Person()
                person("FirstName") = "Louis"
                person("LastName") = "Dejardin"
                Assert.AreEqual("Louis", person.FirstName)
                Assert.AreEqual("Dejardin", person.LastName)
            End If
            If True Then
                Dim person As Object = [New].Person(FirstName:="Bertrand", LastName:="Le Roy").Aliases("bleroy", "boudin")

                Assert.AreEqual("Bertrand", person.FirstName)
                Assert.AreEqual("Le Roy", person.LastName)
                Assert.AreEqual("boudin", Impromptu.InvokeGetIndex(person.Aliases, 1))
            End If

            If True Then
                Dim person = [New].Person().FirstName("Louis").LastName("Dejardin").Aliases(New String() {"Lou"})

                Assert.AreEqual("Louis", person.FirstName)
                Assert.AreEqual("Lou", Impromptu.InvokeGetIndex(person.Aliases, 0))
            End If

            If True Then
                Dim person = [New].Person(New With { _
                 Key .FirstName = "Louis", _
                 Key .LastName = "Dejardin" _
                })
                Assert.AreEqual("Louis", person.FirstName)
                Assert.AreEqual("Dejardin", person.LastName)
            End If

        End Sub

        <Test()> _
        Public Sub TestBuilderActLikeAnon()
            Dim [New] = Builder.[New]().ActLike(Of IBuilder)()

            Dim tNest = [New].Nester(New With { _
             Key .NameLevel1 = "Lvl1", _
             Key .Nested = [New].Nester2(New With { _
              Key .NameLevel2 = "Lvl2" _
             }) _
            })

            Assert.AreEqual("Lvl1", tNest.NameLevel1)
            Assert.AreEqual("Lvl2", tNest.Nested.NameLevel2)
        End Sub

        <Test()> _
        Public Sub TestBuilderActLikeNamed()
            Dim [New] = Builder.[New]().ActLike(Of IBuilder)()

            Dim tNest = [New].Nester(NameLevel1:="Lvl1", Nested:=[New].Nester2(NameLevel2:="Lvl2"))

            Assert.AreEqual("Lvl1", tNest.NameLevel1)
            Assert.AreEqual("Lvl2", tNest.Nested.NameLevel2)
        End Sub


        'This test data is modified from MS-PL Clay project http://clay.codeplex.com
        <Test()> _
        Public Sub TestFactoryListSyntax()
            Dim [New] As Object = Builder.[New]()

            'Test using Clay Syntax
            Dim people = [New].Array([New].Person().FirstName("Louis").LastName("Dejardin"), [New].Person().FirstName("Bertrand").LastName("Le Roy"))

            Assert.AreEqual("Dejardin", people(0).LastName)
            Assert.AreEqual("Le Roy", people(1).LastName)

            Dim people2 = New ImpromptuList() From { _
             [New].Robot(Name:="Bender"), _
             [New].Robot(Name:="RobotDevil") _
            }


            Assert.AreEqual("Bender", people2(0).Name)
            Assert.AreEqual("RobotDevil", people2(1).Name)

        End Sub

        <Test()> _
        Public Sub TestQuicListSyntax()
            Dim tList = Build.NewList("test", "one", "two")
            Assert.AreEqual("one", tList(1))
        End Sub
        Private Shared Function InlineAssignHelper(Of T)(ByRef target As T, value As T) As T
            target = value
            Return value
        End Function

    End Class
End Namespace

File: Tests/UnitTestImpromptuInterface/ImpromptuDynamic.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Dynamic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using ImpromptuInterface;
using ImpromptuInterface.Dynamic;

#if !SELFRUNNER
using DynObj = Dynamitey.DynamicObjects;

using NUnit.Framework;
#endif


#if SILVERLIGHT
namespace UnitTestImpromptuInterface.Silverlight
#else
namespace UnitTestImpromptuInterface
#endif
{
    [TestFixture]
    public class ImpromptuDynamic : Helper
    {

        [Test]
        public void DictionaryPropertyTest()
        {

            dynamic tNew = new ImpromptuDictionary();
            tNew.Prop1 = "Test";
            tNew.Prop2 = 42L;
            tNew.Prop3 = Guid.NewGuid();

            var tActsLike = Impromptu.ActLike<ISimpeleClassProps>(tNew);

            Assert.AreEqual(tNew.Prop1, tActsLike.Prop1);
            Assert.AreEqual(tNew.Prop2, tActsLike.Prop2);
            Assert.AreEqual(tNew.Prop3, tActsLike.Prop3);
        }

        [Test]
        public void DictionaryNullPropertyTest()
        {

            dynamic tNew = new ImpromptuDictionary();


            ISimpeleClassProps tActsLike = Impromptu.ActLike<ISimpeleClassProps>(tNew);

            Assert.AreEqual(default(string), tActsLike.Prop1);
            Assert.AreEqual(default(long), tActsLike.Prop2);
            Assert.AreEqual(default(Guid), tActsLike.Prop3);
        }

        [Test]
        public void GetterAnonTest()
        {
            var tAnon = new { Prop1 = "Test", Prop2 = 42L, Prop3 = Guid.NewGuid() };

            dynamic tTest =new ImpromptuGet(tAnon);

            Assert.AreEqual(tAnon.Prop1, tTest.Prop1);
            Assert.AreEqual(tAnon.Prop2, tTest.Prop2);
            Assert.AreEqual(tAnon.Prop3, tTest.Prop3);
        }

        [Test]
        public void GetterVoidTest()
        {
            var tPoco = new VoidMethodPoco();

            dynamic tTest = new ImpromptuGet(tPoco);

            tTest.Action();
        }

        [Test]
        public void GetterArrayTest()
        {
		
			
            var tArray = new int[]{1,2,3};

            IStringIntIndexer tTest =  ImpromptuGet.Create<IStringIntIndexer>(tArray);

            Assert.AreEqual(tArray[2].ToString(), tTest[2]);
        }

        [Test]
        public void GetterEventTest()
        {
            var tActsLike = ImpromptuGet.Create<IEvent>(new PocoEvent());
            var tSet = false;
            tActsLike.Event += (obj, args) => tSet = true;

            tActsLike.OnEvent(null, null);
            Assert.AreEqual(true, tSet);

        }


        [Test]
        public void GetterEventTest2()
        {
            var tActsLike = ImpromptuGet.Create<IEvent>(new PocoEvent());
            var tSet = false;
            EventHandler<EventArgs> tActsLikeOnEvent = (obj, args) => tSet = true;
            tActsLike.Event += tActsLikeOnEvent;
            tActsLike.Event -= tActsLikeOnEvent;
            tActsLike.OnEvent(null, null);
            Assert.AreEqual(false, tSet);

        }


        [Test]
        public void GetterDynamicTest()
        {
            dynamic tNew = new ExpandoObject();
            tNew.Prop1 = "Test";
            tNew.Prop2 = 42L;
            tNew.Prop3 = Guid.NewGuid();

            dynamic tTest = new ImpromptuGet(tNew);


            Assert.AreEqual(tNew.Prop1, tTest.Prop1);
            Assert.AreEqual(tNew.Prop2, tTest.Prop2);
            Assert.AreEqual(tNew.Prop3, tTest.Prop3);
        }

        [Test]
        public void ForwardAnonTest()
        {
            var tAnon = new { Prop1 = "Test", Prop2 = 42L, Prop3 = Guid.NewGuid() };

            dynamic tTest = new TestForwarder(tAnon);

            Assert.AreEqual(tAnon.Prop1, tTest.Prop1);
            Assert.AreEqual(tAnon.Prop2, tTest.Prop2);
            Assert.AreEqual(tAnon.Prop3, tTest.Prop3);
        }

        [Test]
        public void ForwardVoidTest()
        {
            var tPoco = new VoidMethodPoco();

            dynamic tTest = new TestForwarder(tPoco);

            tTest.Action();
        }



        [Test]
        public void ForwardDynamicTest()
        {
            dynamic tNew = new ExpandoObject();
            tNew.Prop1 = "Test";
            tNew.Prop2 = 42L;
            tNew.Prop3 = Guid.NewGuid();

            dynamic tTest = new TestForwarder(tNew);


            Assert.AreEqual(tNew.Prop1, tTest.Prop1);
            Assert.AreEqual(tNew.Prop2, tTest.Prop2);
            Assert.AreEqual(tNew.Prop3, tTest.Prop3);
        }

        [Test]
        public void DictionaryMethodsTest()
        {

            dynamic tNew = new ImpromptuDictionary();
            tNew.Action1 = new Action(Assert.Fail);
            tNew.Action2 = new Action<bool>(Assert.IsFalse);
            tNew.Action3 = new Func<string>(() => "test");
            tNew.Action4 = new Func<int, string>(arg => "test" + arg);


            ISimpeleClassMeth2 tActsLike = Impromptu.ActLike<ISimpeleClassMeth2>(tNew);



            AssertException<AssertionException>(tActsLike.Action1);
            AssertException<AssertionException>(() => tActsLike.Action2(true));

            Assert.AreEqual("test", tActsLike.Action3());

            Assert.AreEqual("test4", tActsLike.Action4(4));
        }

        [Test]
        public void ForwardMethodsTest()
        {

            dynamic tNew = new ImpromptuDictionary();
            tNew.Action1 = new Action(Assert.Fail);
            tNew.Action2 = new Action<bool>(Assert.IsFalse);
            tNew.Action3 = new Func<string>(() => "test");
            tNew.Action4 = new Func<int, string>(arg => "test" + arg);


            dynamic tFwd = new TestForwarder(tNew);



            AssertException<AssertionException>(()=> tFwd.Action1());
            AssertException<AssertionException>(() => tFwd.Action2(true));

            Assert.AreEqual("test", tFwd.Action3());

            Assert.AreEqual("test4", tFwd.Action4(4));
        }

        [Test]
        public void DictionaryMethodsOutTest()
        {

            dynamic tNew = new ImpromptuDictionary();
            tNew.Func = new DynamicTryString(TestOut);

            String tOut;

            Assert.AreEqual(true, tNew.Func(null, "Test", out tOut));
            Assert.AreEqual("Test", tOut);

            String tOut2;
            Assert.AreEqual(false, tNew.Func(null, 1, out tOut2));
            Assert.AreEqual(null, tOut2);
        }

        private static object TestOut(CallSite dummy, object @in, out string @out)
        {
            @out = @in as string;

            return @out != null;
        }


        [Test]
        public void DictionaryMethodsTestWithPropertyAccess()
        {

            dynamic tNew = new ImpromptuDictionary();
            tNew.PropCat = "Cat-";
            tNew.Action1 = new Action(Assert.Fail);
            tNew.Action2 = new Action<bool>(Assert.IsFalse);
            tNew.Action3 = new ThisFunc<string>(@this => @this.PropCat + "test"); 

            ISimpeleClassMeth tActsLike = Impromptu.ActLike<ISimpeleClassMeth>(tNew);



            AssertException<AssertionException>(tActsLike.Action1);
            AssertException<AssertionException>(() => tActsLike.Action2(true));

            Assert.AreEqual("Cat-test", tActsLike.Action3());


        }

        [Test]
        public void DictionaryNullMethodsTest()
        {

            dynamic tNew = new ImpromptuDictionary();

            ISimpleStringMethod tActsLike = Impromptu.ActLike<ISimpleStringMethod>(tNew);

            Assert.AreEqual(false, tActsLike.StartsWith("Te"));

 

        }
        #if !SELFRUNNER
        [Test]
        public void DynamiteyDictionaryNullMethodsTest()
        {

            dynamic tNew = new DynObj.Dictionary();

            ISimpleStringMethod tActsLike = Impromptu.ActLike<ISimpleStringMethod>(tNew);

            Assert.AreEqual(false, tActsLike.StartsWith("Te"));



        }
    #endif

        [Test]
        public void DynamicDictionaryWrappedTest()
        {

            var tDictionary = new Dictionary<string, object>
                                  {
                                      {"Test1", 1},
                                      {"Test2", 2},
                                      {"TestD",  new Dictionary<string,object>()
                                                     {
                                                         {"TestA","A"},
                                                         {"TestB","B"}
                                                     }
                                      }
                                  };

            dynamic tNew = new ImpromptuDictionary(tDictionary);

            Assert.AreEqual(1, tNew.Test1);
            Assert.AreEqual(2, tNew.Test2);
            Assert.AreEqual("A", tNew.TestD.TestA);
            Assert.AreEqual("B", tNew.TestD.TestB);
        }

        [Test]
        public void InterfaceDictionaryWrappedTest()
        {

            var tDictionary = new Dictionary<string, object>
                                  {
                                      {"Test1", 1},
                                      {"Test2", 2L},
                                      {"Test3",1},
                                      {"Test4","Two"},
                                      {"TestD",  new Dictionary<string,object>()
                                                     {
                                                         {"TestA","A"},
                                                         {"TestB","B"}
                                                     }
                                      }
                                  };

            dynamic tDynamic = ImpromptuDictionary.Create<IDynamicDict>(tDictionary);
            dynamic tNotDynamic = ImpromptuDictionary.Create<INonDynamicDict>(tDictionary);

            Assert.AreEqual(tDynamic, tNotDynamic);

            Assert.AreEqual(1, tDynamic.Test1);
            Assert.AreEqual(2L, tDynamic.Test2);
            Assert.AreEqual(TestEnum.One, tDynamic.Test3);
            Assert.AreEqual(TestEnum.Two, tDynamic.Test4);

            Assert.AreEqual("A", tDynamic.TestD.TestA);
            Assert.AreEqual("B", tDynamic.TestD.TestB);

            Assert.AreEqual(1, tNotDynamic.Test1);
            Assert.AreEqual(2L, tNotDynamic.Test2);
            Assert.AreEqual(TestEnum.One, tNotDynamic.Test3);
            Assert.AreEqual(TestEnum.Two, tNotDynamic.Test4);

            Assert.AreEqual(typeof(Dictionary<string, object>), tNotDynamic.TestD.GetType());
            Assert.AreEqual(typeof(ImpromptuDictionary), tDynamic.TestD.GetType());
        }

        [Test]
        public void DynamicObjectEqualsTest()
        {
            var tDictionary = new Dictionary<string, object>
                                  {
                                      {"Test1", 1},
                                      {"Test2", 2},
                                      {"TestD",  new Dictionary<string,object>()
                                                     {
                                                         {"TestA","A"},
                                                         {"TestB","B"}
                                                     }
                                      }
                                  };

            dynamic tDynamic = ImpromptuDictionary.Create<IDynamicDict>(tDictionary);
            dynamic tNotDynamic = ImpromptuDictionary.Create<INonDynamicDict>(tDictionary);

            Assert.AreEqual(tDynamic, tNotDynamic);

            Assert.AreEqual(tDynamic, tDictionary);

            Assert.AreEqual(tNotDynamic, tDictionary);
        }

        [Test]
        public void DynamicAnnonymousWrapper()
        {
            var tData = new Dictionary<int, string> {{1, "test"}};
            var tDyn = ImpromptuGet.Create(new
                                               {
                                                   Test1 = 1,
                                                   Test2 = "2",
                                                   IsGreaterThan5 = Return<bool>.Arguments<int>(it => it > 5),
                                                   ClearData = ReturnVoid.Arguments(() => tData.Clear())
                                               });

            Assert.AreEqual(1,tDyn.Test1);
            Assert.AreEqual("2", tDyn.Test2);
            Assert.AreEqual(true, tDyn.IsGreaterThan5(6));
            Assert.AreEqual(false, tDyn.IsGreaterThan5(4));

            Assert.AreEqual(1, tData.Count);
            tDyn.ClearData();
            Assert.AreEqual(0, tData.Count);

        }   

        [Test]
        public void TestAnonInterface()
        {
            var tInterface = ImpromptuGet.Create<ICollection>(new
                                                                  {
                                                                     CopyArray = ReturnVoid.Arguments<Array,int>((ar,i) => Enumerable.Range(1,10)),
                                                                     Count =  10,
                                                                     IsSynchronized = false,
                                                                     SyncRoot = this,
                                                                     GetEnumerator = Return<IEnumerator>.Arguments(()=>Enumerable.Range(1, 10).GetEnumerator())
                                                                  });

            Assert.AreEqual(10, tInterface.Count);
            Assert.AreEqual(false, tInterface.IsSynchronized);
            Assert.AreEqual(this, tInterface.SyncRoot);
            Assert.AreEqual(true,tInterface.GetEnumerator().MoveNext());
        }
		  
		[Test]
		public void TestBuilder(){
			var New = Builder.New<ExpandoObject>();
			
			  var tExpando =New.Object(
										Test:"test1",
										Test2:"Test 2nd"
									);
		   Assert.AreEqual("test1",tExpando.Test );
		   Assert.AreEqual("Test 2nd",tExpando.Test2);
			
		    dynamic NewD = new ImpromptuBuilder<ExpandoObject>();

			
			var tExpandoNamedTest =NewD.Robot(
				LeftArm:"Rise",
				RightArm:"Clamp"
				);
			
			 Assert.AreEqual("Rise",tExpandoNamedTest.LeftArm);
		  	 Assert.AreEqual("Clamp",tExpandoNamedTest.RightArm);
		}
				
		[Test]
		public void TestSetupOtherTypes(){
			var New = Builder.New().Setup(
					Expando: typeof(ExpandoObject),
					Dict: typeof(ImpromptuDictionary)
				);
			
			var tExpando =New.Expando(
				LeftArm:"Rise",
				RightArm:"Clamp"
				);
			
			var tDict =New.Dict(
				LeftArm:"RiseD",
				RightArm:"ClampD"
				);
			
			 Assert.AreEqual("Rise",tExpando.LeftArm);
		  	 Assert.AreEqual("Clamp",tExpando.RightArm);
			Assert.AreEqual(typeof(ExpandoObject),tExpando.GetType());
			
			 Assert.AreEqual("RiseD",tDict.LeftArm);
			 Assert.AreEqual("ClampD",tDict.RightArm);
			Assert.AreEqual(typeof(ImpromptuDictionary),tDict.GetType());

		}
		
		[Test]

        //This test data is modified from MS-PL Clay project http://clay.codeplex.com
        public void TestClayFactorySyntax()
        {
            dynamic New = Builder.New();

            {
                var person = New.Person();
                person.FirstName = "Louis";
                person.LastName = "Dejardin";
                Assert.AreEqual("Louis",person.FirstName );
                Assert.AreEqual("Dejardin", person.LastName);
            }
            {
                var person = New.Person();
                person["FirstName"] = "Louis";
                person["LastName"] = "Dejardin";
                Assert.AreEqual("Louis", person.FirstName);
                Assert.AreEqual("Dejardin", person.LastName);
            }
            {
                var person = New.Person(
                    FirstName: "Bertrand",
                    LastName: "Le Roy"
                    ).Aliases("bleroy", "boudin");

                Assert.AreEqual("Bertrand", person.FirstName);
                Assert.AreEqual("Le Roy", person.LastName);
                Assert.AreEqual("boudin", person.Aliases[1]);
            }

            {
                var person = New.Person()
                    .FirstName("Louis")
                    .LastName("Dejardin")
                    .Aliases(new[] {"Lou"});

                Assert.AreEqual(person.FirstName, "Louis");
                Assert.AreEqual(person.Aliases[0], "Lou");
            }

            {
                var person = New.Person(new
                {
                    FirstName = "Louis",
                    LastName = "Dejardin"
                });
                Assert.AreEqual(person.FirstName, "Louis");
                Assert.AreEqual(person.LastName, "Dejardin");
            }

        }
		
		[Test]
		public void TestBuilderActLikeAnon()
		{
		    var New = Builder.New().ActLike<IBuilder>();

		var tNest =New.Nester(new {
				NameLevel1 = "Lvl1",
				Nested =  New.Nester2(new{
					NameLevel2 = "Lvl2"
				})
			});
			
			Assert.AreEqual("Lvl1", tNest.NameLevel1);
			Assert.AreEqual("Lvl2",tNest.Nested.NameLevel2);
	    }

        [Test]
        public void TestBuilderActLikeNamed()
        {
            var New = Builder.New().ActLike<IBuilder>();

            var tNest = New.Nester(
                NameLevel1 :"Lvl1",
                Nested : New.Nester2(
                            NameLevel2 : "Lvl2"
                        )
            );

            Assert.AreEqual("Lvl1", tNest.NameLevel1);
            Assert.AreEqual("Lvl2", tNest.Nested.NameLevel2);
        }
		
		
			[Test]
        //This test data is modified from MS-PL Clay project http://clay.codeplex.com
		public void TestFactoryListSyntax(){
			dynamic New = Builder.New();
			
			//Test using Clay Syntax
			var people = New.Array(
    						 New.Person().FirstName("Louis").LastName("Dejardin"),
   							 New.Person().FirstName("Bertrand").LastName("Le Roy")
						 );
			
			Assert.AreEqual("Dejardin",people[0].LastName);
		  	Assert.AreEqual("Le Roy",people[1].LastName);
			
			var people2 =  new ImpromptuList(){
							New.Robot(Name:"Bender"),
							New.Robot(Name:"RobotDevil")
						   };	
			
			
			Assert.AreEqual("Bender", people2[0].Name);
		  	Assert.AreEqual("RobotDevil",people2[1].Name);

		}
              
        [Test]
            public void TestQuicListSyntax()
            {
                var tList =Build.NewList("test", "one", "two");
                Assert.AreEqual("one", tList[1]);

                var tList2 = Build.NewList("test", "one", "two", "three");
                Assert.AreEqual("three", tList2[3]);
            }

        
        [Test]
        public void TestLateLibrarybind()
        {

            dynamic tBigIntType = new ImpromptuLateLibraryType("System.Numerics.BigInteger, System.Numerics, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089");

            if (tBigIntType.IsAvailable)
            {

                var one = tBigIntType.@new(1);
                var two = tBigIntType.@new(2);

                Assert.IsFalse(one.IsEven);
                Assert.AreEqual(true, two.IsEven);

                var tParsed =tBigIntType.Parse("4");

                Assert.AreEqual(true, tParsed.IsEven);



            }
            else
            {
#if !SILVERLIGHT 
                Assert.Fail("Big Int Didn't Load");
#endif

            }
        }
		
    }
}
